### PR TITLE
lxd/container: Adds support for nic device host routes

### PR DESCRIPTION
**Pull request by tomponline**
Adds new nic device config keys (comma delimited list of routes):

```
ipv4.routes
ipv6.routes
```

Checklist

- [x] Add support for `ipv4.routes`
- [x] Add support for `ipv6.routes`
- [x] Replace IPv6 disablement with IPv6 RA acceptance disablement
- [x] Add integration tests for adding/removing routes and devices with routes both offline and hotplug
- [x] Add integration tests for devices that don't support routes or limits (macvlan)

### CODE REVIEW

**Comment by tomponline** [comment]
@stgraber this also contains https://github.com/lxc/lxd/pull/5672

**Comment by stgraber** [comment]
Needs:
 - API extension in doc/
 - API extension in shared/version/api.go
 - Doc entry in doc/

**Comment by tomponline** [comment]
Ahh yes, also I need to add check for whether to run on net up hook too

**Comment by freeekanayaka** [comment]
I'm concerned that this broadens the untested surface of ```containerLXC```, and if we are going to rework that code area to introduce support for VMs it will be hard to know if we regress.

**Comment by stgraber** [comment]
Well, that one we can add test for

**Comment by tomponline** [comment]
Will get on it.

File: lxd/container_lxc.go [hunk 1]
```diff
@@ -3013,9 +3017,35 @@ func (c *containerLXC) OnStop(target string) error {
 
 // OnNetworkUp is called by the LXD callhook when the LXC network up script is run.
 func (c *containerLXC) OnNetworkUp(deviceName string, hostName string) error {
+	logger.Debugf("Running network up hook for %s with host name %s", deviceName, hostName)
```
Comment (stgraber) [inline lxd/container_lxc.go]: Can probably be dropped at this point

File: lxd/container_lxc.go [hunk 2]
```diff
@@ -4844,21 +4886,26 @@ func (c *containerLXC) Update(args db.ContainerArgs, userRequested bool) error {
 			if m["type"] == "disk" {
 				updateDiskLimit = true
 			} else if m["type"] == "nic" || m["type"] == "infiniband" {
+				//Check to see if network keys have change
```
Comment (stgraber) [inline lxd/container_lxc.go]: We usually put a space after `//`

File: lxd/container_lxc.go [hunk 3]
```diff
@@ -4844,21 +4886,26 @@ func (c *containerLXC) Update(args db.ContainerArgs, userRequested bool) error {
...
 					if needsUpdate {
 						break
 					}
 				}
 
 				if needsUpdate {
-					// Refresh tc limits
-					m["host_name"] = c.getHostInterface(m["name"])
-					err = c.setNetworkLimits(m)
+					//Populate network device with container nic names.
```
Comment (stgraber) [inline lxd/container_lxc.go]: We usually put a space after `//`

File: lxd/container_lxc.go [hunk 4]
```diff
@@ -4739,6 +4769,18 @@ func (c *containerLXC) Update(args db.ContainerArgs, userRequested bool) error {
 						return err
 					}
 				}
+
+				if m["type"] == "nic" {
+					//Populate network device with container nic names.
```
Comment (stgraber) [inline lxd/container_lxc.go]: We usually put a space after `//`

File: lxd/container_lxc.go [hunk 5]
```diff
@@ -3013,9 +3017,35 @@ func (c *containerLXC) OnStop(target string) error {
...
+}
+
+// SetupHostNetworkDevice configures a nic device's host side settings.
+func (c *containerLXC) SetupHostNetworkDevice(device types.Device) error {
+	//If not already, populate network device with host name.
+	if device["host_name"] == "" {
+		device["host_name"] = c.getHostInterface(device["name"])
+	}
+
+	//Check whether host device resolution succeeded.
```
Comment (stgraber) [inline lxd/container_lxc.go]: We usually put a space after `//`

File: lxd/container_lxc.go [hunk 6]
```diff
@@ -3013,9 +3017,35 @@ func (c *containerLXC) OnStop(target string) error {
...
+	logger.Debugf("Running network up hook for %s with host name %s", deviceName, hostName)
 	device := c.expandedDevices[deviceName]
 	device["host_name"] = hostName
-	return c.setNetworkLimits(device)
+	return c.SetupHostNetworkDevice(device)
+}
+
+// SetupHostNetworkDevice configures a nic device's host side settings.
+func (c *containerLXC) SetupHostNetworkDevice(device types.Device) error {
+	//If not already, populate network device with host name.
```
Comment (stgraber) [inline lxd/container_lxc.go]: We usually put a space after `//`

File: lxd/container_lxc.go [hunk 7]
```diff
@@ -1634,9 +1634,13 @@ func (c *containerLXC) initLXC(config bool) error {
 				}
 			}
 
-			err = lxcSetConfigItem(cc, fmt.Sprintf("%s.%d.script.up", networkKeyPrefix, networkidx), fmt.Sprintf("%s callhook %s %d network-up %s", c.state.OS.ExecPath, shared.VarPath(""), c.id, k))
-			if err != nil {
-				return err
+			//Check if the container has network specific keys set to avoid unnecessarily running the network up hook.
```
Comment (stgraber) [inline lxd/container_lxc.go]: We usually put a space after `//`

File: test/suites/container_devices.sh [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+  fi
+  if ! tc class show dev "${veth_host_name}" | grep "3Mbit" ; then
+    echo "limits.ingress invalid"
+    exit 1
+  fi
+  if ! tc filter show dev "${veth_host_name}" egress | grep "4Mbit" ; then
+    echo "limits.egress invalid"
+    exit 1
+  fi
+}
```
Comment (stgraber) [inline test/suites/container_devices.sh]: Missing trailing carrier return

File: test/suites/container_devices.sh [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+    parent=lxdbr0 \
+    host_name="${veth_host_name}"
+  lxc config device set "${ct_name}" eth0 ipv4.routes "192.0.2.2${ipRand}/32"
+  lxc config device set "${ct_name}" eth0 ipv6.routes "2001:db8::2${ipRand}/128"
+  lxc config device set "${ct_name}" eth0 limits.ingress 3Mbit
+  lxc config device set "${ct_name}" eth0 limits.egress 4Mbit
+
+  if ! ip -4 r | grep "192.0.2.2${ipRand}" ; then
+    echo "ipv4.routes invalid"
+    exit 1
```
Comment (stgraber) [inline test/suites/container_devices.sh]: Do not call `exit` from tests, just call `false`.

Calling exit would bypass all our cleanup code.

Comment (stgraber) [reply test/suites/container_devices.sh]: (Fix all the other ones too :))

**Comment by stgraber** [comment]
jenkins: test this please

**Comment by tomponline** [comment]
thanks @stgraber will get these fixed first thing Tuesday

**Comment by tomponline** [comment]
@stgraber I have fixed those issues now, some of the tests are now passing (since I have improved the isolation of my new tests so as not to interfere with the others). However there are some tests still intermittently failing that appear to be related to DQLITE.

**Comment by stgraber** [comment]
jenkins: test this please

File: lxd/container.go [hunk 10]
```diff
@@ -384,6 +389,32 @@ func containerValidDevices(cluster *db.Cluster, devices types.Devices, profile b
...
 			}
+
+			if m["ipv4.routes"] != "" {
+				if !shared.StringInSlice(m["nictype"], []string{"bridged", "p2p"}) {
+					return fmt.Errorf("Bad nic type for ipv4.routes: %s", m["nictype"])
+				}
+
+				for _, route := range strings.Split(m["ipv4.routes"], ",") {
+					route = strings.TrimSpace(route)
+					if err := networkValidNetworkV4(route); err != nil {
```
Comment (stgraber) [inline lxd/container.go]: Our coding style is to avoid those one-line declaration + check.

File: lxd/container.go [hunk 11]
```diff
@@ -384,6 +389,32 @@ func containerValidDevices(cluster *db.Cluster, devices types.Devices, profile b
...
+			}
+
+			if m["ipv6.routes"] != "" {
+				if !shared.StringInSlice(m["nictype"], []string{"bridged", "p2p"}) {
+					return fmt.Errorf("Bad nic type for ipv6.routes: %s", m["nictype"])
+				}
+
+				for _, route := range strings.Split(m["ipv6.routes"], ",") {
+					route = strings.TrimSpace(route)
+					if err := networkValidNetworkV6(route); err != nil {
```
Comment (stgraber) [inline lxd/container.go]: Our coding style is to avoid those one-line declaration + check.

File: lxd/container_lxc.go [hunk 12]
```diff
@@ -3015,7 +3019,32 @@ func (c *containerLXC) OnStop(target string) error {
 func (c *containerLXC) OnNetworkUp(deviceName string, hostName string) error {
 	device := c.expandedDevices[deviceName]
 	device["host_name"] = hostName
-	return c.setNetworkLimits(device)
+	return c.SetupHostNetworkDevice(device)
+}
+
+// SetupHostNetworkDevice configures a nic device's host side settings.
+func (c *containerLXC) SetupHostNetworkDevice(device types.Device) error {
```
Comment (stgraber) [inline lxd/container_lxc.go]: This is an internal function (not part of the interface) so should be `setupHostNetworkDevice` instead.

File: lxd/container_lxc.go [hunk 13]
```diff
@@ -3015,7 +3019,32 @@ func (c *containerLXC) OnStop(target string) error {
...
+		device["host_name"] = c.getHostInterface(device["name"])
+	}
+
+	// Check whether host device resolution succeeded.
+	if device["host_name"] == "" {
+		return fmt.Errorf("LXC doesn't know about this device and the host_name property isn't set, can't find host side veth name")
+	}
+
+	// Refresh tc limits
+	if err := c.setNetworkLimits(device); err != nil {
```
Comment (stgraber) [inline lxd/container_lxc.go]: Our coding style is to avoid those one-line declaration + check.

File: lxd/container_lxc.go [hunk 14]
```diff
@@ -3015,7 +3019,32 @@ func (c *containerLXC) OnStop(target string) error {
...
+		return err
+	}
+
+	// Setup static routes to container
+	if err := c.setNetworkRoutes(device); err != nil {
```
Comment (stgraber) [inline lxd/container_lxc.go]: Our coding style is to avoid those one-line declaration + check.

File: lxd/container_lxc.go [hunk 15]
```diff
@@ -4739,6 +4768,18 @@ func (c *containerLXC) Update(args db.ContainerArgs, userRequested bool) error {
...
 				}
+
+				if m["type"] == "nic" && shared.StringInSlice(m["nictype"], []string{"bridged", "p2p"}) {
+					// Populate network device with container nic names.
+					m, err := c.fillNetworkDevice(k, m)
+					if err != nil {
+						return err
+					}
+
+					if err = c.SetupHostNetworkDevice(m); err != nil {
```
Comment (stgraber) [inline lxd/container_lxc.go]: Our coding style is to avoid those one-line declaration + check.

File: lxd/container_lxc.go [hunk 16]
```diff
@@ -4844,21 +4885,26 @@ func (c *containerLXC) Update(args db.ContainerArgs, userRequested bool) error {
...
-					m["host_name"] = c.getHostInterface(m["name"])
-					err = c.setNetworkLimits(m)
+				if needsUpdate && shared.StringInSlice(m["nictype"], []string{"bridged", "p2p"}) {
+					// Populate network device with container nic names.
+					m, err := c.fillNetworkDevice(k, m)
 					if err != nil {
 						return err
 					}
+
+					if err = c.SetupHostNetworkDevice(m); err != nil {
```
Comment (stgraber) [inline lxd/container_lxc.go]: Our coding style is to avoid those one-line declaration + check.

File: lxd/container_lxc.go [hunk 17]
```diff
@@ -8586,6 +8632,47 @@ func (c *containerLXC) getHostInterface(name string) string {
...
 }
 
+// setNetworkRoutes applies any static routes configured from the host to the container nic.
+func (c *containerLXC) setNetworkRoutes(m types.Device) error {
+	if !shared.PathExists(fmt.Sprintf("/sys/class/net/%s", m["host_name"])) {
+		return fmt.Errorf("Unknown or missing host side veth: %s", m["host_name"])
+	}
+
+	// Flush all IPv4 routes
+	if _, err := shared.RunCommand("ip", "-4", "route", "flush", "dev", m["host_name"], "proto", "static"); err != nil {
```
Comment (stgraber) [inline lxd/container_lxc.go]: Our coding style is to avoid those one-line declaration + check.

File: lxd/container_lxc.go [hunk 18]
```diff
@@ -8586,6 +8632,47 @@ func (c *containerLXC) getHostInterface(name string) string {
...
+		return err
+	}
+
+	// Flush all IPv6 routes
+	if _, err := shared.RunCommand("ip", "-6", "route", "flush", "dev", m["host_name"], "proto", "static"); err != nil {
```
Comment (stgraber) [inline lxd/container_lxc.go]: Our coding style is to avoid those one-line declaration + check.

**Comment by tomponline** [comment]
@stgraber I'll also add a test for macvlan hotplugging as discussed on https://github.com/lxc/lxd/issues/5686 to make sure this PR doesn't break non-veth device types.

File: test/suites/config.sh [hunk 19]
```diff
@@ -192,13 +192,14 @@ test_config_profiles() {
...
   # test live-adding a nic
+  veth_host_name="veth$$"
   lxc start foo
   lxc exec foo -- cat /proc/self/mountinfo | grep -q "/mnt1.*ro,"
   ! lxc config show foo | grep -q "raw.lxc" || false
   lxc config show foo --expanded | grep -q "raw.lxc"
   ! lxc config show foo | grep -v "volatile.eth0" | grep -q "eth0" || false
   lxc config show foo --expanded | grep -v "volatile.eth0" | grep -q "eth0"
-  lxc config device add foo eth2 nic nictype=p2p name=eth10
+  lxc config device add foo eth2 nic nictype=p2p name=eth10 host_name="${veth_host_name}"
```
Comment (stgraber) [inline test/suites/config.sh]: This should be its own commit I think as that would need backporting along with the NetworkUp hook.

File: lxd/container_lxc.go [hunk 20]
```diff
@@ -1634,9 +1634,13 @@ func (c *containerLXC) initLXC(config bool) error {
 				}
 			}
 
-			err = lxcSetConfigItem(cc, fmt.Sprintf("%s.%d.script.up", networkKeyPrefix, networkidx), fmt.Sprintf("%s callhook %s %d network-up %s", c.state.OS.ExecPath, shared.VarPath(""), c.id, k))
-			if err != nil {
-				return err
+			// Check if the container has network specific keys set to avoid unnecessarily running the network up hook.
+			containerNetworkKeys := append(containerNetworkLimitKeys, containerNetworkRouteKeys...)
+			if shared.StringMapHasStringKey(m, containerNetworkKeys...) && shared.StringInSlice(m["nictype"], []string{"bridged", "p2p"}) {
```
Comment (stgraber) [inline lxd/container_lxc.go]: Can you split this into its own earlier commit so we can cherry-pick it into stable?

That is, the earlier commit shouldn't mention containerNetworkRouteKeys, then this commit should just alter the logic to also cover containerNetworkRouteKeys.

File: test/suites/config.sh (hunk 19 above)
Comment (tomponline) [reply test/suites/config.sh]: ok will do @stgraber. Are you happy (perhaps wrong word), that the network up hook will break hotplugging p2p/bridge devices when kernel doesn't have the `netnsid_getifaddrs` kernel feature? This is because in the past, when a veth device was hotplugged, it's tc limits were not applied (and nor will its static routes be applied). 

In principle, we could only block the the hotplug from working if one or more of limits/routes were defined on the nic, and allow a simple device that doesn't need host side setup to complete on earlier kernels.

**Comment by tomponline** [comment]
@stgraber macvlan tests added now using dummy host side interface.

File: lxd/container_lxc.go (hunk 20 above)
Comment (tomponline) [reply lxd/container_lxc.go]: @stgraber this is now done

File: test/suites/config.sh (hunk 19 above)
Comment (tomponline) [reply test/suites/config.sh]: @stgraber this is now done

Comment (tomponline) [reply test/suites/config.sh]: @stgraber as discussed the check for network host side keys has been added so that you can still hotplug a nic on an older kernel if you are not using those features. This was not needed for the limits only feature because in the current stable branch the limits keys are not inspected/applied when hotplugging a new nic anyway.
